snare

```diff
@@ -0,0 +1,72 @@
+#! /usr/bin/env python3.7
+
+# Copyright (c) 2019 King's College London, created by the Software Development
+# Team <http://soft-dev.org/>
+#
+# Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+# http://www.apache.org/licenses/LICENSE-2.0>, or the MIT license <LICENSE-MIT
+# or http://opensource.org/licenses/MIT>, or the UPL-1.0 license
+# <http://opensource.org/licenses/UPL> at your option. This file may not be
+# copied, modified, or distributed except according to those terms.
+
+import getopt, hashlib, hmac, json, os, subprocess, sys, tempfile, traceback
+from http.server import ThreadingHTTPServer, BaseHTTPRequestHandler
+from urllib.parse import unquote
+
+class HookHandler(BaseHTTPRequestHandler):
+    def do_POST(self):
+        data = self.rfile.read(int(self.headers["Content-Length"]))
+        if not self._check_secret(data):
+            self.send_response(401)
+            return
+
+        data_str = unquote(data.decode("utf-8")[8:])
+        pl = json.loads(data_str)
+        owner = pl["repository"]["owner"]["login"].lower()
+        repo = pl["repository"]["name"].lower()
+
+        path = os.path.join(repo_scripts_dir, "%s_%s" % (owner, repo))
+        if os.path.exists(path):
+            with tempfile.NamedTemporaryFile() as jsonf:
+                jsonf.write(data_str.encode())
+                event = self.headers["X-GitHub-Event"]
+                rtn = subprocess.run([path, event, jsonf.name])
+            if rtn.returncode == 0:
+                self.send_response(200)
+            else:
+                self.send_response(500)
+        else:
+            self.send_response(501)
+
+    def _check_secret(self, data):
+        sha, sig = self.headers["X-Hub-Signature"].split("=")
+        mac = hmac.new(secret.encode(), msg=data, digestmod=hashlib.sha1)
+        return hmac.compare_digest(mac.hexdigest(), sig)
+
+def usage(code):
+    sys.stderr.write("Usage: snare -p <port> -r <repo scripts dir> -s <secret>\n")
+    sys.exit(code)
+
+if __name__ == '__main__':
+    port = None
+    repo_scripts_dir = None
+    secret = None
+    opts, args = getopt.getopt(sys.argv[1 : ], "hp:r:s:")
+    for opt in opts:
+        if opt[0] == "-h":
+            usage(0)
+        elif opt[0] == "-p":
+            port = int(opt[1])
+        elif opt[0] == "-r":
+            repo_scripts_dir = os.path.abspath(opt[1])
+        elif opt[0] == "-s":
+            secret = opt[1]
+    if port is None or repo_scripts_dir is None or secret is None or len(args) > 0:
+        usage(1)
+
+    while True:
+        try:
+            server = ThreadingHTTPServer(('', port), HookHandler)
+            server.serve_forever()
+        except Exception:
```

Comment:
I recommend you don't catch such a generic exception. I used to do this in krun, and it hides bugs and makes debugging a nightmare.

Reply:
In a server, you've got to keep serving no matter what, so there isn't much choice.

Reply:
I understand that you want to keep going if there's some temporary network error, but some classes of exception you are going to want to know about.

Eg `IndexError`, `KeyError`, `ValueError`. These would all indicate bugs in our script, and your catch all would silence them. We'd never know that our hooks are not executing.

Reply:
Then we will crash, and we won't get a bug report either. We can't really win, so it's generally better for the server to keep running.

Reply:
don't you at least want to print the exception before continuing?

Reply:
It does.

Reply:
Based on my experience with Krun, I think this decision will come back to haunt you. Ultimately it's up to you, since you are maintaining it ;)

Another option would be to notify someone (via email?) if X hooks have failed.

Come to think about it, wouldn't you want to know if any single hook failed? Otherwise our docs will be out of date. Do we need some kind of retry logic then?

Reply:
At the moment everything gets logged to stderr/stdout. I'm wary of doing anything more clever until I understand a bit better what the failure modes are.

Reply:
OK. So you are going to manually check stderr/stdout to start with? Perhaps run in a tmux?

Reply:
Yes, in a tmux.

Reply:
Sorry to bang on, but where do you print to stderr/stderr? It looks like you catch any exception and silently discard it.

Wouldn't you want:
```
except Exception as e:
    print(e)
```

?

Reply:
The HTTP server prints the exception.

Reply:
Ah I see!

Reply:
Reading the `serve_forever()` docs, it's not clear if the server will re-raise exceptions. But I take it you added the handler because it was?

It looks like it boils down to a call to:
https://docs.python.org/2/library/socketserver.html#SocketServer.BaseServer.handle_error

(I promise I will leave you alone soon)

Reply:
It works as is even when things go a bit wrong. Perfect? Probably not. Adequate? For now, yes.

Reply:
Point taken.